allenact/embodiedai/aux_losses/losses.py

```diff
@@ -546,6 +551,192 @@ def get_aux_loss(
+        )  # (T+k-1, N, -1)
+
+        ## unfold function will create consecutive action sequences
+        action_seq = (
+            action_padded.unfold(dimension=0, size=self.planning_steps, step=1)
+            .permute(3, 0, 1, 2)
+            .view(self.planning_steps, num_steps * num_samplers, action_embed_size)
+        )  # (k, T*N, -1)
+
+        ## beliefs GRU output
+        obs_embeds = aux_model.visual_mlp(obs_embeds)  # (T, N, 128)
+
+        beliefs = beliefs.view(1, num_steps * num_samplers, -1)  # (1, T*N, -1)
+
+        # get future contexts c_{t+1:t+k} = GRU(b_t, a_{t:t+k-1})
+        future_contexts_all, _ = aux_model.context_model(
+            action_seq, beliefs
+        )  # (k, T*N, -1)
+
+        future_contexts_all = aux_model.belief_mlp(future_contexts_all)  # (k, T*N, 128)
+        future_contexts_all = future_contexts_all.view(-1, 128)  # (k*T*N, 128)
+
+        obs_embeds = obs_embeds.view(
+            num_steps * num_samplers, obs_embeds.shape[-1]
+        ).permute(
+            1, 0
+        )  # (-1, T*N)
+
+        visual_logits = torch.matmul(future_contexts_all, obs_embeds)
+        visual_log_probs = F.log_softmax(visual_logits, dim=1)  ## (k*T*N, T*N)
+
+        target = torch.zeros(
+            (self.planning_steps, num_steps, num_samplers),
+            dtype=torch.long,
+            device=beliefs.device,
+        )  # (k, T, N)
+        loss_mask = torch.zeros(
+            (self.planning_steps, num_steps, num_samplers), device=beliefs.device
+        )  # (k, T, N)
+
+        num_valid_before = 0
+        for j in range(num_samplers):
+            for i in range(num_steps):
+                index = i * num_samplers + j
+
+                if i == 0 or masks[i, j].item() == 0:
+                    num_valid_before = 0
+                    continue
+
+                num_valid_before += 1
+                for back in range(min(num_valid_before, self.planning_steps)):
+                    target[back, i - (back + 1), j] = index
+                    loss_mask[back, i - (back + 1), j] = 1.0
+
+        target = target.view(-1)  # (k*T*N,)
+
+        loss_value = self.cross_entropy_loss(visual_log_probs, target)
+        loss_value = loss_value.view(
+            self.planning_steps, num_steps, num_samplers, 1
+        )  # (k, T, N, 1)
+
+        loss_mask = loss_mask.unsqueeze(-1)  # (k, T, N, 1)
+        loss_valid_masks = loss_mask * _bernoulli_subsample_mask_like(
+            loss_mask, self.subsample_rate
+        )  # (k, T, N, 1)
+
+        num_valid_losses = torch.count_nonzero(loss_valid_masks)
+
+        avg_multi_class_loss = (loss_value * loss_valid_masks).sum() / torch.clamp(
+            num_valid_losses, min=1.0
+        )
+
+        return (
+            avg_multi_class_loss,
+            {"total": cast(torch.Tensor, avg_multi_class_loss).item(),},
+        )
+
+
+######## CPCA Softmax variants ######
+
+
+class CPCA1SoftMaxLoss(CPCASoftMaxLoss):
```

Comment:
It's a bit tedious to have so many similar classes just because of the UUID. Would it be possible to have the uuid be an argument to `__init__` and have some sort of factory that gives us the right instance given the needed uuid? Didn't think thoroughly about the cons of such a design, so feel free to ignore if it's completely out of scope.

Reply:
Interesting! I could even imagine that the uuid could be something like `"CPCASoftMaxLoss(planning_steps=8)"` and then we could do some reflection to instantiate the class. For now I'll merge but will keep this in mind for the future.